### PR TITLE
TFP: asinh pressure scale ablation (0.75, 0.5) — isolation test

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Asinh pressure normalization (`--asinh-pressure`) has not been tested in isolation on the TandemFoil Paper dataset. The noam baseline (PR #3025) folds `--asinh-pressure` into the full physics stack (`--enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --residual-prediction --enable-pressure-prior-addition`), making it impossible to attribute how much asinh contributes alone. Historical data shows asinh was -8% on noam's OOD metrics in an earlier ablation, but the open question is whether it independently helps TFP's `val_primary/field_mse`.

This experiment tests two asinh scale values (0.75 and 0.5) against the TFP champion config with `--enable-fourier` only — stripping the full physics stack to isolate the asinh effect. This fills the only missing cell in the asinh isolation matrix: emma (#3189) tests asinh on TF, casca (#3192) tests asinh on DM, but nobody tests asinh alone on TFP.

## Instructions

Use the TFP champion config throughout (Lion lr=1.25e-4, T_max=10, gc=0.5, WD=1e-2, EMA=0.999, 3L/192d/3H, Fourier enabled). Do **not** add `--enable-te-coord-frame`, `--enable-cp-panel`, `--enable-cp-panel-tandem-only`, `--residual-prediction`, or `--enable-pressure-prior-addition` — this is a clean single-variable ablation.

**Step 0 — 3-epoch smoke test (run first):**

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --ema-decay 0.999 \
  --asinh-pressure --asinh-scale 0.75 \
  --epochs 3 \
  --wandb_group tfp-asinh-ablation
```

Report `val_primary/field_mse` at epoch 3. If it diverges or is clearly worse than ~0.010, stop and comment before proceeding.

**Trial 1 — asinh-scale 0.75 (full run):**

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --ema-decay 0.999 \
  --asinh-pressure --asinh-scale 0.75 \
  --epochs 999 \
  --wandb_group tfp-asinh-ablation
```

**Trial 2 — asinh-scale 0.5 (full run):**

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --ema-decay 0.999 \
  --asinh-pressure --asinh-scale 0.5 \
  --epochs 999 \
  --wandb_group tfp-asinh-ablation
```

Report best `val_primary/field_mse` for each trial (from best-validation checkpoint, not terminal epoch). Include W&B run IDs for both.

## Baseline

Current best metrics from BASELINE.md (PR #3025 — haku/tfp-lion-champion-config):

| Metric | Value |
|--------|-------|
| val_primary/field_mse | **0.002383** (epoch 443) |
| val/surface_mse | 0.001517 |
| val/surface_mse_p | 4.81e-05 |
| val/volume_mse | 0.002397 |

- Baseline W&B run: d1xh0o1p (haku/tfp-lion-champion-config)
- Reproduce command:
```bash
cd target/icml2026 && python train.py --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999
```

Target to beat: `val_primary/field_mse < 0.002383`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT gohan:

## Smoke Test Results (Step 0 — 3-epoch, asinh-scale 0.75)

| Metric | Epoch 1 | Epoch 2 | Epoch 3 |
|--------|---------|---------|---------|
| val_primary/field_mse | 6.38e16 | 4.21e13 | 5.89e7 |
| val/surface_mse_p | 1.01e12 | 8.23e9 | 1.37e7 |
| val/surface_mse_Ux | 0.114 | 0.065 | 0.025 |
| val/surface_mse_Uy | 0.075 | 0.070 | 0.066 |
| loss | 0.858 | 0.506 | 0.392 |

**W&B run:** iut6s3oo (gohan/tfp-asinh-075-smoke)

**Assessment:** val_primary/field_mse at epoch 3 is 5.89e7 — clearly above the ~0.010 threshold. However, the model is **not diverging** — it's converging very steeply, improving by ~10^9x over just 3 epochs. The velocity components (Ux, Uy) are in a reasonable range; the pressure component is driving all the error.

This is likely a sinh denormalization amplification effect: asinh compresses the training targets, but when computing metrics via sinh(predicted) the model's early-epoch pressure errors get exponentially amplified. With the full physics stack (residual prediction, pressure prior addition, Cp panels), the pressure channel has much better initialization and guidance. Without those features, the model starts from scratch on pressure prediction in asinh-space, and sinh amplifies the early errors into huge denormalized values.

**Convergence rate:** field_mse dropped by ~10^9x in 3 epochs. If the superexponential convergence continues, it could reach ~0.01 levels within ~10-15 more epochs, then enter a slower optimization phase.

**Recommendation:** I believe the 3-epoch snapshot is uninformative for asinh normalization due to the sinh amplification effect. The strong convergence trend suggests the model will eventually reach a meaningful range. However, per your instructions, I'm stopping to ask: should I proceed with the full trials, or is this indicative of a fundamental issue with asinh in isolation on TFP?

**Comment by tcapelle** [comment]
STUDENT gohan:

## Results

### Trial 1 — asinh-scale 0.75

| Metric | Value | vs Baseline |
|--------|-------|-------------|
| val_primary/field_mse | **0.775** (epoch 324) | 325x worse (baseline 0.002383) |
| test_primary/field_mse | 0.654 | 274x worse |
| test/surface_mse | 0.754 | |
| test/surface_mse_p | 0.645 | |
| test/volume_mse | 0.652 | |
| Epochs trained | 556 (360-min timeout) | |

- **W&B run:** 08s8cxns (gohan/tfp-asinh-075-full)
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 --ema-decay 0.999 --asinh-pressure --asinh-scale 0.75 --epochs 999 --wandb_group tfp-asinh-ablation
```

**Stability:** Two divergence events (~ep275 and ~ep465) where loss spiked from ~0.04 to >1.0. The best checkpoint (ep324, field_mse=0.775) was found during chaotic recovery from the first divergence — not through stable optimization. The model was never stably in a competitive regime.

### Trial 2 — asinh-scale 0.5

| Metric | Value | vs Baseline |
|--------|-------|-------------|
| val_primary/field_mse | **34,048** (epoch 501) | ~14 million x worse |
| test_primary/field_mse | 10,934,985 | |
| test/surface_mse | 698,475,200 | |
| test/surface_mse_p | 2,095,425,536 | |
| test/volume_mse | 320 | |
| Epochs trained | 543 (360-min timeout) | |

- **W&B run:** pqawyn4k (gohan/tfp-asinh-050-full)
- **Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 --ema-decay 0.999 --asinh-pressure --asinh-scale 0.5 --epochs 999 --wandb_group tfp-asinh-ablation
```

**Stability:** No divergence in training loss (steadily decreased from 0.89 to 0.016), but denormalized field_mse oscillated wildly (1e5 to 1e12 range) and never converged. Best checkpoint at ep501 was still 34,048 — ~14 million times above baseline.

## What happened

**Both trials are clear negatives.** Asinh pressure normalization in isolation (Fourier-only, no physics stack) does not work on TFP. Neither scale value (0.75 or 0.5) produced competitive field_mse.

**Root cause:** The asinh/sinh transform creates a fundamental amplification problem on denormalized metrics. During training, the model learns in asinh-compressed space where loss decreases normally (both runs showed healthy loss convergence: 0.8→0.04). But when denormalizing via sinh() for metric evaluation, any residual pressure prediction error gets exponentially amplified. Without the physics stack (residual prediction, pressure prior, Cp panels), the model has no mechanism to constrain pressure predictions into the narrow asinh-space range where sinh denormalization produces reasonable physical values.

**Why the baseline works:** The baseline (PR #3025) uses asinh WITH the full physics stack. Residual prediction gives the model a better initialization by predicting corrections to freestream, and the pressure prior addition provides a physics-informed anchor for pressure. These constrain the pressure channel enough that asinh normalization can help rather than hurt.

**Trial 1 vs Trial 2:** Scale 0.75 is slightly less aggressive than 0.5, and it managed a lucky 0.775 best during post-divergence chaos. Scale 0.5 compresses pressure more aggressively, making the sinh amplification problem worse — hence the much higher (34K vs 0.775) best field_mse. However, Trial 1's 0.775 is unreliable since it came from a chaotic state, not stable optimization.

## Bug fix

Fixed an `UnboundLocalError` in `train.py` where a local variable `primary_metric_key` (string) at line 1646 shadowed the function `primary_metric_key()` at line 1428, causing the function call at line 1639 to fail. Renamed the local to `checkpoint_metric_key`.

## Suggested follow-ups

1. **Asinh + residual prediction only** — test whether residual prediction alone is sufficient to stabilize asinh on TFP (without the full physics stack).
2. **Asinh + pressure prior addition only** — same question for pressure prior.
3. **This fills the asinh isolation matrix** — emma (#3189) tests TF, casca (#3192) tests DM, and now this PR establishes that asinh alone fails on TFP. The conclusion is consistent: asinh needs the physics stack for support.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3198 because ADVISOR: Closing — both trials catastrophically worse than baseline. Trial 1 (asinh-0.75): field_mse=0.775 apparent best but 60/554 steps were Inf (chaotic transient, not stable minimum). Trial 2 (asinh-0.50): field_mse=34,047 — 14M x worse. Root cause: without residual prediction and pressure prior, the model predicts in asinh-compressed space but sinh() denormalization at eval exponentially amplifies residual pressure errors. Velocity channels healthy — failure is entirely pressure-channel-driven. Note: this is NOT the cp_panel code regression — these runs don't use the physics stack. Conclusive: asinh alone is incompatible with TFP. Bug fix for primary_metric_key shadowing noted. Good diagnostic work, gohan — redirecting you to AF volume closure..
